go-controller/pkg/ovn/policy.go

```diff
@@ -1098,29 +1130,45 @@ func (oc *Controller) buildNetworkPolicyACLs(np *networkPolicy, aclLogging strin
 	return acls
 }
 
-func (oc *Controller) deleteNetworkPolicy(policy *knet.NetworkPolicy) {
+// deleteNetworkPolicy removes a network policy
+// If np is provided, then deletion may still occur without a lock on nsInfo
+func (oc *Controller) deleteNetworkPolicy(policy *knet.NetworkPolicy, np *networkPolicy) error {
 	klog.Infof("Deleting network policy %s in namespace %s",
 		policy.Name, policy.Namespace)
 
+	var foundNp *networkPolicy
 	nsInfo, nsUnlock := oc.getNamespaceLocked(policy.Namespace, false)
 	if nsInfo == nil {
-		klog.V(5).Infof("Failed to get namespace lock when deleting policy %s in namespace %s",
-			policy.Name, policy.Namespace)
-		return
+		// if we didn't get nsInfo and np is nil, we cannot proceed
+		if np == nil {
+			return fmt.Errorf("failed to get namespace lock when deleting policy %s in namespace %s",
+				policy.Name, policy.Namespace)
+		}
+	} else {
+		defer nsUnlock()
+		foundNp = nsInfo.networkPolicies[policy.Name]
 	}
-	defer nsUnlock()
 
-	np := nsInfo.networkPolicies[policy.Name]
-	if np == nil {
-		return
+	// try to use the more official np found in nsInfo
+	if foundNp != nil {
+		np = foundNp
+	}
+
+	if err := oc.destroyNetworkPolicy(np, nsInfo); err != nil {
```

Comment:
you can reach here with nsInfo==nil and np!=nil case, I think your nested if's broke the logic of the older code. Is that on purpose?

Reply:
the intention with this change was to be able to delete without nsInfo, because it is possible that the namespace could be gone by the time we want to delete a network policy. Here is a case where that may happen:

```
	// Now do nsinfo operations to set the policy
	nsInfo, nsUnlock, err = oc.ensureNamespaceLocked(policy.Namespace, false, nil)
	if err != nil {
		// rollback network policy
		if err := oc.deleteNetworkPolicy(policy, np); err != nil {
			// rollback failed, add to retry to cleanup
			oc.addDeleteToRetryPolicy(policy, np)
		}
		return fmt.Errorf("unable to ensure namespace for network policy: %s, namespace: %s, error: %v",
			policy.Name, policy.Namespace, err)
	}
```